File: 1. Algorithmic Toolbox/Week3/33_dot_product.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using std::vector;

long long max_dot_product(vector<int> a, vector<int> b) {
	std::sort(begin(a), end(a)); std::sort(begin(b), end(b));
	int size = a.size();
	long long sum = 0;				
	for (int i = 0; i < size; i++) {
		sum += (long long) a[i]*b[i];
	}		
	return sum;
}

int main() {
  size_t n;
  std::cin >> n;
  vector<int> a(n), b(n);
  for (size_t i = 0; i < n; i++) {
    std::cin >> a[i];
  }
  for (size_t i = 0; i < n; i++) {
    std::cin >> b[i];
  }
  std::cout << max_dot_product(a, b) << std::endl;
}

File: 1. Algorithmic Toolbox/Week3/33_DotProduct.java
import java.util.Scanner;
import java.util.Arrays;

class DotProduct {

    private static long maxDotProduct(final int[] a, final int[] b) {
		
		Arrays.sort(a); Arrays.sort(b);
		int size = a.length;
		int sum = 0;				
		for (int i = 0; i < size; i++) {
			sum += a[i]*b[i];
		}		
		return sum;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
            a[i] = scanner.nextInt();
        }
        int[] b = new int[n];
        for (int i = 0; i < n; i++) {
            b[i] = scanner.nextInt();
        }
        System.out.println(maxDotProduct(a, b));
    }
}

